[TOPIC-GPIO] drivers: hci: spi: update to new GPIO API

erwango opened the pull request:
Update bluetooth hcp spi driver to new GPIO API.

Tested on disco_l475_iot1 and 96b_carbon.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

mnkp commented on boards/arm/96b_carbon/96b_carbon.dts:
```diff
@@ -84,8 +84,8 @@
 		reg = <0>;
-		reset-gpios = <&gpiob 4 0>;
-		irq-gpios = <&gpiob 1 0>;
+		reset-gpios = <&gpiob 4 GPIO_ACTIVE_HIGH>;
```
Since the driver is using logical levels to drive the reset signal I believe we should have here `GPIO_ACTIVE_LOW` as the reset is activated when the line is low. The driver code would need to be updated to set value `1` to activate the reset. Otherwise - this is an open point, haven't been discussed so far - we could also have the driver use raw / physical signals. In this case it should use `gpio_pin_set_raw` function to set the pin value and we shouldn't have here `GPIO_ACTIVE_HIGH` flag.

erwango commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -479,17 +478,18 @@ static int bt_spi_send(struct net_buf *buf)
+
+	gpio_pin_interrupt_configure(irq_dev, GPIO_IRQ_PIN,
+				     GPIO_INT_EDGE_RISING);
 
```
Thanks, so I'm moving the `init/add_callback` function before `gpio_pin_interrupt_configure` then.

erwango commented:
@mnkp, updated, thanks for the review

erwango commented:
@mnkp, updated

mnkp commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -480,26 +482,21 @@ static int bt_spi_open(void)
-		return -EINVAL;
-	}
+	gpio_pin_interrupt_configure(irq_dev, GPIO_IRQ_PIN,
+				     GPIO_INT_EDGE_RISING);
```
Since we want to support logical levels (honor `GPIO_ACTIVE_LOW`, `GPIO_ACTIVE_HIGH` flags) we should use here `GPIO_INT_EDGE_TO_ACTIVE` flag.

mnkp commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -201,22 +204,21 @@ static int configure_cs(void)
-			   GPIO_DIR_OUT | GPIO_PUD_PULL_UP);
-	gpio_pin_write(cs_dev, GPIO_CS_PIN, 1);
+	gpio_pin_configure(cs_dev, GPIO_CS_PIN, GPIO_OUTPUT | GPIO_CS_FLAGS);
+	gpio_pin_set(cs_dev, GPIO_CS_PIN, 1);
```
This PR may need to be rebased but latest GPIO API supports `GPIO_OUTPUT_ACTIVE` flag. The two calls can be replaced with
```
	gpio_pin_configure(cs_dev, GPIO_CS_PIN, GPIO_OUTPUT_ACTIVE | GPIO_CS_FLAGS);
```

mnkp commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -480,26 +482,21 @@ static int bt_spi_open(void)
-			   GPIO_DIR_OUT | GPIO_PUD_PULL_UP);
-	gpio_pin_write(rst_dev, GPIO_RESET_PIN, 0);
+			   GPIO_OUTPUT | GPIO_RESET_FLAGS);
+	gpio_pin_set(rst_dev, GPIO_RESET_PIN, 1);
```
`gpio_pin_set` can be removed if `GPIO_OUTPUT_ACTIVE` flag is used.

mnkp reviewed (commented):
LGTM with minor comments.

erwango commented:
@mnkp, updated

mnkp commented on boards/arm/disco_l475_iot1/disco_l475_iot1.dts:
```diff
@@ -118,13 +118,13 @@
 	status = "okay";
 
-	cs-gpios = <&gpiod 13 0>, <&gpioe 0 0>;
+	cs-gpios = <&gpiod 13 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>, <&gpioe 0 0>;
```
I overlooked it earlier, do we need to add flags also to `gpioe`? Also, this change doesn't seem to be related to hci-spi driver. Maybe we should do it in another PR?

erwango replied:
`gpioe` is dealt with in https://github.com/zephyrproject-rtos/zephyr/pull/21066.

I see it related in the fact that it deals with the same global functionality which is hci BLE  over SPI. So I have been able to test and validate porting to new API in current change. I don't see much value in splitting it.

erwango committed:
drivers: hci: spi: update to new GPIO API

Update bluetooth hcp spi driver to new GPIO API.
Following changes have been done:
- Use new gpio api functions
- Get reset, irq and cs flags from dt defines
- Define reset pin as active low and invert set/unset logic.

Tested on disco_l475_iot1 and 96b_carbon.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
